README.md
[![pre-commit](https://img.shields.io/badge/pre--commit-enabled-brightgreen?logo=pre-commit)](https://github.com/pre-commit/pre-commit)

Healthcheck Monitoring Dashboard : Using this application we can easily monitor all the API or Service healthcheck and keep track of their status . This application will monitor preodically based on preconfigured time intervals in config file . Dashboard will show the status check mark with clour code and status code message . In case of any error we can view the Error message .

## Archtecture :

Dashboard --> (ReactJS) --> (NodeJS) --> Target API's \

```bash
"proxy": http://localhost:3838
```

![docker view](Dashboard.png)
![docker view](ErrorMessage.png)

## Getting Started with Healthcheck Monitoring

---

```bash
npm install

npm run build

Runs the app [ node server.js / nodemon server.js / pm2 start server.js ]\
Open [http://localhost:3838](http://localhost:3838) to view it in the browser.
or, Direct React UI using [http://localhost:3001](http://localhost:3001)
```

---

## Setting up .env

```bash
chmod 777 start.sh
./start.sh
or,
touch .env
STATUS_SERVER_URL=http://localhost:3838
```

## Running pre-commit checks

[pre-commit](https://pre-commit.com) installs git hooks configured in [.pre-commit-config.yaml](.pre-commit-config.yaml)

Install `pre-commit` and `commitizen` to use

---

```bash

brew install commitizen
brew install pre-commit

pre-commit install
pre-commit install --hook-type commit-msg

pre-commit run --all-files
```

---

## Update package.json with the latest version of all dependencies

```bash
npm install -g npm-check-updates
ncu -u
npm install

```


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

### 0.1.2 (2023-06-01)


### Features

* update packages ([f8ebfff](https://github.com/jayantapaul-18/healthcheck-monitoring/commit/f8ebfff761ef23bac076c4974d9011c3cc80f53f))


server.js
//process.env["NODE_TLS_REJECT_UNAUTHORIZED"] = 0;
const express = require('express');
const path = require('path');
const fs = require('fs');
const app = express();
const process = require("process");
const helmet = require("helmet");
const bodyParser = require("body-parser");
const cookieParser = require("cookie-parser");
const cors = require("cors");
const { v4: uuidv4 } = require('uuid');
const status = require("./status.js");
const newUUID = uuidv4();
const { createProxyMiddleware } = require('http-proxy-middleware');

// app.use(helmet());
// app.use('/', createProxyMiddleware({ target: 'http://localhost:3838', changeOrigin: true }));
app.use(helmet.hidePoweredBy());
app.use(bodyParser.urlencoded({ extended: false }));
app.use(bodyParser.json());
// app.use(bodyParser.json({ type: "application/json" }));
app.use(cookieParser());
// Render using Prod Build
app.use(express.static(path.join(__dirname, 'build')))
// Render from Public Directory
// app.use(express.static(path.join(__dirname, 'public')))
app.use(cors());
app.use("/app/v1", status);

//allow cross origin requests
app.use(function (req, res, next) {
    res.setHeader(
      "Access-Control-Allow-Methods",
      "GET, POST, PUT, OPTIONS, DELETE"
    );
    res.header("Access-Control-Allow-Origin", ["http://localhost:3000","http://localhost:3838","*"]);
    res.header(
      "Access-Control-Allow-Headers",
      "Origin, X-Requested-With, Content-Type, Accept"
    );
    res.header("Access-Control-Allow-Credentials", false);
    next();
  });


app.get('/app/v1/ping', (req, res) => {
    return res.send({ status:'pong'})
})

app.get('/', (req, res) => {
  res.sendFile(path.join(__dirname, 'build', 'index.html'))
})
app.get('/target-url-config', (req, res) => {
    res.sendFile(path.join(__dirname, 'build', 'index.html'))
  })


function startTimer() {
    const started = process.uptime()
    return () => Math.floor((process.uptime() - started) * 1000) // uptime gives us seconds
  }

app.all("/app/v1/healthz", (req, res) => {
    // winston.info(`${req.method} - ${req.originalUrl} - ${req.ip}`);
    const rss = (process.memoryUsage().rss / (1024 * 1024)).toFixed(2);
    const heapTotal = (process.memoryUsage().heapTotal / (1024 * 1024)).toFixed(2);
    const uptime = (process.uptime() / 60).toFixed(2);
    res.status(200).send({ message: "healthy", status: 200, application: "api", upTime: uptime ,heapTotal:heapTotal,rss:rss});
  });



// const options = {
//   key: fs.readFileSync('test/fixtures/keys/agent2-key.pem'),
//   cert: fs.readFileSync('test/fixtures/keys/agent2-cert.pem')
// };

function startServer() {
    const port =
      process.env.NODE_ENV === "production" ? process.env.TOOL_SERVER_PORT || 80 : 3838;
    app.listen(port, () => {
      console.log(
        "Server listening on port http://localhost:" +
          port +
          "/app/v1/healthz  [ " + newUUID +" ]"+
          " [ How to Message in a World of Noise ]"
      );
      // winston.info(`Server listening on port http://localhost:${port}`);
    });
  }

startServer();
// @End //


status.js
const status = require("express")();
const os = require('os');
const v8 = require('v8');
const eventLoopStats = require('event-loop-stats');
const axios = require('axios');
const moment = require('moment');
const { TARGET_URL , TIMEOUT} = require('./config');



status.get("/target-url", (req, res) => {
    var newArray = TARGET_URL.map((object, i) => ({ ...object, key: i + 1 }));
    res.status(200).json(newArray);
});

status.get("/status", (req, res) => {

    var status = {
        timestamp: Date.now(),
        osHostname: os.hostname(),
        processArch: process.arch,
        osPlatform:os.platform(),
        heapStat: v8.getHeapStatistics(),
        osFreemem: os.freemem(),
        osTotalmem :os.totalmem(),
        osLoadavg: os.loadavg(),
        osCpus: os.cpus(),
        osNetworkInterfaces: os.networkInterfaces(),
        osUptime: os.uptime(),
        v8GetHeapSpaceStatistics:v8.getHeapSpaceStatistics()

    }

    res.status(200).json({status:status, eventLoopStats:eventLoopStats.sense()});
  });

// message: '"connect ETIMEDOUT ip:443"'
// message: '"Request failed with status code 404"'
// message: '"timeout of 1000ms exceeded"'

function allSettled (promises) {
    const allPromises = promises.map(p => Promise.resolve(p)
        .then(
        value => ({ state: 'fulfilled', value: value }),
        error => ({ state: 'rejected', reason: error })

      )
    );
    return Promise.all(allPromises);
  }


status.get("/health-status", (req, res) => {
    const Promises = [];
    const TARGET_TEST_RESULTS = [];
    (TARGET_URL || []).forEach(TARGET => {
        let uri = `${TARGET.protocol}://${TARGET.hostPort}${TARGET.path}`;
        Promises.push(axios({
          url: uri,
          method: 'GET',
            timeout: TIMEOUT,
            maxRedirects: 2
        },
        ));
      });

    var startDate = moment();
    allSettled(Promises).then(results => {
        results.forEach((result, index) => {
            var endDate = moment();
            console.log('Request took: ' + endDate.diff(startDate) + ' ms.');
            // if(result.status == 'fulfilled' && result.value.status == 200) {
            //     console.log(result.value.status);
            //   }
            //   if(result.status == 'fulfilled' && result.value.status == 502) {
            //     console.log(result.value.status);
            //   }
            if (result.state === 'rejected') {
                console.log("rejected result: ",result);
                console.log(JSON.stringify(result.reason.message));
                TARGET_TEST_RESULTS.push({
                    key:index,
                    path: TARGET_URL[index].path,
                    status: ['failed'],
                    httpStatusCode: (result.reason.code),
                    responseBody: JSON.stringify(result),
                    response: "Error: "+JSON.stringify(result.reason.message),
                    timestamp: new Date
                });
            } else {
                // console.log(result);
                TARGET_TEST_RESULTS.push({
                    key:index,
                    path: TARGET_URL[index].path,
                    status: ['success'],
                    httpStatusCode: 200,
                    responseBody: '',
                    response: result.value.data,
                    timestamp: new Date
                });
            }
        });

        //   return TARGET_TEST_RESULTS;
        console.log(TARGET_TEST_RESULTS);
        res.status(200).json(TARGET_TEST_RESULTS);
    });

});

module.exports = status;


config.js
const dotenv = require('dotenv');
dotenv.config();

module.exports = {
    TIMEOUT: 5000,
    REQUEST_INTERVAL: 1000
}

module.exports = {
    TARGET_URL : [
        {
            key: 1,
            tag: 'plab',
            protocol: 'http',
            hostPort: 'localhost:3838',
            path: '/app/v1/ping',
            status: ['success'],
            httpStatusCode: 200,
            responseBody: '',
            response: { message: 33.589 },
            timestamp: "2022-03-26T15:56:10.493Z",
            state: true
        },
        {
            tag: 'plab',
            protocol: 'http',
            hostPort: 'localhost:3838',
            path: '/app/v1/healthz',
            state: true
        },
        {
            tag: 'plab',
            protocol: 'http',
            hostPort: 'localhost:1881',
            path: '/app/v1/healthcheck',
            state: true
        },
        {
            tag: 'plab',
            protocol: 'http',
            hostPort: 'localhost:3003',
            path: '/secure/v1/healthcheck',
            state: false
        },
        {
            tag: 'plab',
            protocol: 'https',
            hostPort: 'localhost:3003',
            path: '/yup/v1/healthcheck',
            state: false
        },
        {
            tag: 'plab',
            protocol: 'http',
            hostPort: '192.168.1.173:1881',
            path: '/agent/v1/healthcheck',
            state: false
        },
        {
            tag: 'plab',
            protocol: 'http',
            hostPort: 'localhost:1881',
            path: '/perf/v1/healthcheck',
            state: false
        },
        {
            tag: 'plab',
            protocol: 'http',
            hostPort: '192.168.1.173:1881',
            path: '/agent/v1/getTemp',
            state: false
        },
        {
            tag: 'plab',
            protocol: 'http',
            hostPort: 'localhost:1881',
            path: '/agent/v1/cpuLoad',
            state: false
        },
        {
            tag: 'plab',
            protocol: 'http',
            hostPort: '192.168.1.173:1881',
            path: '/agent/v1/getEnv',
            state: false
        },
        {
            tag: 'plab',
            protocol: 'http',
            hostPort: '192.168.1.173:1881',
            path: '/agent/v1/hostnamectl',
            state: false
        },
        {
            tag: 'plab',
            protocol: 'http',
            hostPort: '192.168.1.173:1881',
            path: '/agent/v1/getTemp',
            state: false
        }
    ]

}
